crates/gas/src/call_graph/node.rs

```diff
@@ -0,0 +1,246 @@
+/// * References to other Nodes it has edges connected to.
+///
+pub struct Node<T> {
+    value: T,
+    incoming: HashSet<NodeRef<T>>,
+    outgoing: HashSet<NodeRef<T>>,
+}
+
+impl<T> Node<T>
+where
+    T: Value,
+{
+    pub fn new(value: T) -> Self {
+        Self {
+            value,
+            incoming: HashSet::new(),
+            outgoing: HashSet::new(),
+        }
+    }
+
+    /// Returns the value associated with the `Node`
+    pub fn value(&self) -> T {
+        self.value
+    }
+
+    /// Returns the a `Vec` of references to `Node`(s) with incoming edges to self
+    pub fn incoming(&self) -> Vec<NodeRef<T>> {
+        self.incoming.iter().cloned().collect()
+    }
+
+    /// Whether the current node has incoming edges
+    pub fn has_incoming(&self) -> bool {
+        self.incoming.len() > 0
+    }
+
+    /// Returns the a `Vec` of references to outgoing `Node`(s) from `self`
+    pub fn outgoing(&self) -> Vec<NodeRef<T>> {
+        self.outgoing.iter().cloned().collect()
+    }
+
+    /// Whether the current node has outgoing edges
+    pub fn has_outgoing(&self) -> bool {
+        self.outgoing.len() > 0
+    }
+
+    /// Adds an outgoing edge to `dest` (i.e: `self` -> `dest`)
+    pub fn add_out_edge(&mut self, dest: NodeRef<T>) {
+        self.outgoing.insert(dest);
+    }
+
+    /// Adds an incoming edge from `dest` (i.e: `dest` -> `self`)
+    pub fn add_in_edge(&mut self, source: NodeRef<T>) {
+        self.incoming.insert(source);
+    }
+
+    /// Removes an incoming edge from `dest` (i.e: `dest` -> `self`)
+    pub fn remove_in_edge(&mut self, source: &NodeRef<T>) {
+        self.incoming.remove(source);
+    }
+
+    /// Removes an outgoing edge to `dest` (i.e: `self` -> `dest`)
+    pub fn remove_out_edge(&mut self, dest: &NodeRef<T>) {
+        self.outgoing.remove(dest);
+    }
+
+    /// Whether is a `sink` node. (i.e: has no outgoing edges)
+    pub fn is_sink(&self) -> bool {
+        self.outgoing.is_empty()
+    }
+
+    /// Whether is a `source` node. (i.e: has no incoming edges)
+    pub fn is_source(&self) -> bool {
+        self.incoming.is_empty()
+    }
+
+    /// Whether is an `isolated` node. (i.e: has neither incoming nor outgoing edges)
+    pub fn is_isolated(&self) -> bool {
+        (self.is_sink() == false) && (self.is_source() == false)
+    }
+}
+
+impl<T> Hash for Node<T>
+where
+    T: Value,
+{
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.value().hash(state);
+    }
+}
+
+impl<T> PartialEq for Node<T>
+where
+    T: Value,
+{
+    fn eq(&self, other: &Self) -> bool {
+        let value = self.value();
+        let other = other.value();
+
+        value.eq(&other)
+    }
+}
+
+impl<T> Eq for Node<T> where T: Value {}
+
+/// The `NodeRef wrapper struct it adds readability to the code
+/// (less cognitive load instead of seeing `Rc<RefCell<Node<T>>>` all over the place
+#[repr(transparent)]
+pub struct NodeRef<T> {
+    inner: Rc<RefCell<Node<T>>>,
```

Comment:
Can you elaborate a bit on the need for using `Rc<RefCell<Node<T>>>`?

Reply:
It's a common practice in Rust:
https://doc.rust-lang.org/book/ch15-05-interior-mutability.html#having-multiple-owners-of-mutable-data-by-combining-rct-and-refcellt